### ISSUE TITLE
Use more concrete exception types in `test_builtin`

### ISSUE THREAD

**Issue by sobolevn** [python/cpython#99124]
There are several cases where more general exception cases are used instead of concrete ones.

I think it is important to use the very specific ones because otherwise we leave a possible space for contract changes.

For example:

```python
        with self.assertRaises(ValueError):
            A.__name__ = 'A\udcdcB'
```

is not 100% correct, because it actually raises: `UnicodeEncodeError` which is a not-direct subtype of `ValueError`.

* PR: gh-99125

* PR: gh-99224

* PR: gh-99225

**Comment by iritkatriel** [comment]
> is not 100% correct, because it actually raises: `UnicodeEncodeError` which is a not-direct subtype of `ValueError`.
> 

The question is what the documentation says it raises. If it's documented to raise ValueError then the fact that it actually raises a subclass of ValueError is an implementation detail, and should not be asserted in the tests.

**Comment by sobolevn** [comment]
@iritkatriel indeed, let's look into it.

But, I think that `builtins` are quite special: they are so widely used that every implementation detail is actually a part of the contract.

People are expecting this code to work right now:

```python
try:
   __import__('something')
except ModuleNotFoundError:
    # some error logic
```

So, changing it for some other `ValueError` subclass is not just an implementation detail change, it is a breaking change (in any case). And in my opinion, it must be asserted.

But, I am still interested about what docs say.

## `__import__`

Current docs do not say anything about any exceptions at all: https://docs.python.org/3/library/functions.html#import__

So, there are no promises from us, except the behaviour we already have.

## `__name__`

There might be some other parts where this is referenced. But, now I can only find https://docs.python.org/3/library/stdtypes.html#definition.__name__ and https://docs.python.org/3/library/exceptions.html?highlight=unicodedecodeerror#UnicodeDecodeError

The second link says:

> Raised when a Unicode-related error occurs during decoding. It is a subclass of [UnicodeError](https://docs.python.org/3/library/exceptions.html?highlight=unicodedecodeerror#UnicodeError).

And https://docs.python.org/3/library/functions.html#type

> This is essentially a dynamic form of the [class](https://docs.python.org/3/reference/compound_stmts.html#class) statement. The name string is the class name and becomes the [__name__](https://docs.python.org/3/library/stdtypes.html#definition.__name__) attribute.

So, basically my tests follows this logic: 
- we create a new `type()` with `__name__`
- unicode cannot be decoded
- `UnicodeDecodeError` is raised

It matches the docs we have, and is quite consitent with other parts of CPython.
